### PR TITLE
Single Sign-On Authentication Mechanism

### PR DESCRIPTION
**Pull request by TheMysteriousX**
This is an authentication module that can be used to integrate LibreNMS with single sign-on providers.

The authentication code is imperative rather than object oriented; this has meant that I've had to make more changes than I'd like.

There are tests included, but they must be executed standalone and with config.php manually set (hence the .disabled extension) - the authentication code being imperative means that once a mechanism is loaded, it cannot be unloaded or replaced, breaking other tests.

I've also stopped http-auth being special cased in authenticate.php by tweaking it to use the two methods I've added.

It's been a few years since I've written any PHP, so feedback is gratefully received.

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

### CODE REVIEW

**Comment by TheMysteriousX** [comment]
Looks like there are a few compatibility problems to fix up - I'll sort those momentarily.

**Comment by TheMysteriousX** [comment]
Code now passes Travis' checks under PHP 5.3.
I can't do much about the Scrutinizer failures; what it's flagging up is 100% correct, but is due to the auth mechanisms having repeated code and identical functions in each file.

**Comment by laf** [comment]
Thanks for contributing this.

I've not gone through the code in detail yet and have no real way to test this but it seems like it would be better to put the config into another level within the config array like radius: https://docs.librenms.org/#Extensions/Authentication/#radius-authentication

I.e `$config['sso']['sso_group_strategy']`

**Comment by TheMysteriousX** [comment]
Just pushed a new version with the configuration moved down a level.

I've not tested it this way, but it should be possible to simulate a single sign-on environment using SetEnv - something like this in your Apache config:

```
SetEnv REMOTE_USER ssouser
SetEnv mail ssouser@example.org
SetEnv displayName "SSO User"
SetEnv level 10
```

Then a basic config like:

```
$config['sso']['mode'] = 'env';
$config['sso']['realname_attr'] = 'displayName';
$config['sso']['email_attr'] = 'mail';
$config['sso']['level_attr'] = 'level';
$config['sso']['create_users'] = true;
$config['sso']['update_users'] = true;
$config['sso']['group_strategy'] = 'attribute';
```

Which should create an entry in the db for 'ssouser', and allow access at level 10. Might be a few rough edges in this configuration, but it should work.

File: html/includes/authentication/sso.inc.php [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+ * @author     Adam Bishop <adam@omega.org.uk>
+ */
+
+use LibreNMS\Config;
+use LibreNMS\Util\IP;
+use LibreNMS\Exceptions\AuthenticationException;
+use LibreNMS\Exceptions\InvalidIpException;
+
+/**
+ * Some functionality in this mechanism is inspired by confluence_http_authenticator (@chauth) and graylog-plugin-auth-sso (@Graylog)
```
Comment (laf) [inline html/includes/authentication/sso.inc.php]: Please add both of these packages to the Acknowledgements doc: https://github.com/librenms/librenms/blob/master/doc/General/Acknowledgement.md

File: html/includes/authentication/sso.inc.php [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+
+function init_auth()
+{
+    // No initialisation is required
+}
+
+function authenticate($username, $password)
+{
+    if (empty($username)) {
+        throw new AuthenticationException('sso.user_attr was not found or was empty');
```
Comment (laf) [inline html/includes/authentication/sso.inc.php]: Users don't see config values like that so you will need to `$config['sso']['user_attr']` in this output

File: html/includes/authentication/sso.inc.php [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+    // Build the user's details from attributes
+    $email = auth_sso_get_attr(Config::get('sso.email_attr'));
+    $realname = auth_sso_get_attr(Config::get('sso.realname_attr'));
+    $description = auth_sso_get_attr(Config::get('sso.descr_attr'));
+    $can_modify_passwd = 0;
+
+    $level = auth_sso_calculate_level();
+
+    // User has already been approved by the authenicator so if automatic user create/update is enabled, do it
+    if (Config::get('sso.create_users') && !user_exists($username)) {
```
Comment (laf) [inline html/includes/authentication/sso.inc.php]: Is this block going to be called on every page and every graph load as we run `authenticate()` each time.

File: html/includes/print-menubar.php [hunk 4]
```diff
@@ -639,7 +639,7 @@
 
 if ($_SESSION['authenticated']) {
     echo('
-           <li><a href="logout/"><i class="fa fa-sign-out fa-fw fa-lg" aria-hidden="true"></i> Logout</a></li>');
+           <li><a href="' . $config['auth_logout_handler']  . '"><i class="fa fa-sign-out fa-fw fa-lg" aria-hidden="true"></i> Logout</a></li>');
```
Comment (laf) [inline html/includes/print-menubar.php]: Should switch to `Config::get()` here as well.

File: html/includes/authentication/sso.inc.php (hunk 1 above)
Comment (TheMysteriousX) [reply html/includes/authentication/sso.inc.php]: Done.

File: html/includes/authentication/sso.inc.php (hunk 3 above)
Comment (TheMysteriousX) [reply html/includes/authentication/sso.inc.php]: I don't believe that's the case - I threw in a print statement and authenticate() only seems to be called on the first request. It gets short circuited if there's an active session.

If it did get called repeatedly (user with broken cookie settings) it's idempotent, and if a database cache is present, the impact is just an additional sql query per page load.

File: html/includes/print-menubar.php (hunk 4 above)
Comment (TheMysteriousX) [reply html/includes/print-menubar.php]: Done, by changing how this is done. The template isn't really the right place to handle authentication stuff as it's prone to being altered. I implemented the same functionality by making the URL the user is sent to after log out configurable.

File: html/includes/authentication/sso.inc.php (hunk 2 above)
Comment (TheMysteriousX) [reply html/includes/authentication/sso.inc.php]: Done.

File: html/includes/authentication/sso.inc.php (hunk 3 above)
Comment (laf) [reply html/includes/authentication/sso.inc.php]: Actually I've just remember we switched to using sessions once validated so you're right on this.

**Comment by laf** [comment]
LGTM.

@librenms/reviewers anyone else?

**Comment by laf** [comment]
one thing to note actually. Someone has got a PR in the move all of the auth modules to classes #7497 

As that's been in for a while, it would be unfair to ask that user to do the work to convert this as well. Would you be able to migrate this once 7497 is merged?

**Comment by TheMysteriousX** [comment]
Sure that's fine - I have a couple of questions about the new structure, I'll pop them in the other pull request.

**Comment by TheMysteriousX** [comment]
Answered my questions with a bit more reading.

**Comment by laf** [comment]
@TheMysteriousX The other PR is now merged in.

**Comment by TheMysteriousX** [comment]
Naïve rebase, do not merge as it probably doesn't work (yet).

**Comment by TheMysteriousX** [comment]
I think this is now ported. I'll give it another once over with fresher eyes tomorrow.

Summary of changes:
* Migrated a couple of methods into the interface/base class
* Converted imperative mechanism to class based
* Enabled tests as authentication is now mutable
* Added Auth::reset to allow a new mechanism to be loaded at runtime

The CI check that is failing is PHP5.3 - I've used the wrong array syntax to support this.

I can go back and change the syntax, but I think PHP5.3 was being dropped?

**Comment by laf** [comment]
It's not being dropped until January so this will need to be resolved before then.

File: LibreNMS/Authentication/AuthorizerBase.php [hunk 5]
```diff
@@ -244,4 +245,14 @@ public function updateUser($user_id, $realname, $level, $can_modify_passwd, $ema
         //not supported by default
         return 0;
     }
+
+    public function authExternal()
```
Comment (murrant) [inline LibreNMS/Authentication/AuthorizerBase.php]: This would probably be better named `authIsExternal()`

**Comment by murrant** [comment]
:+1: for implementing unit tests :)

**Comment by TheMysteriousX** [comment]
Great, thanks for reviewing.

**Comment by TheMysteriousX** [comment]
Thanks for merging - the company I work for use this mechanism on their instance, so feel free to @ me if there are any issues.

**Comment by murrant** [comment]
@TheMysteriousX I'm looking to refactor this a bit to inheret from MysqlAuthorizer.

I can't figure out why you ignore the given $username in functions and use $this->getExternalUsername() instead?  Would this cause errors if we query for another user besides the logged in one?

**Comment by TheMysteriousX** [comment]
That would be fine.

The username passed by functions should always be taken by from getExternalUsername, so by taking it from there it eliminates the possibility that the username could have been modified (accidentally or maliciously).

**Comment by murrant** [comment]
But those functions return info about the requested user, not the logged in user. 

For example, say you are viewing a shared dashboard with your code all shared dashboards would show the currently logged in user as the owner.

**Comment by TheMysteriousX** [comment]
Ok, I misunderstood their purpose then - I assumed getUserid/getUserLevel/userExists were intended as internal methods as they're not called outside of the auth code (except for scripts that do weird things, like the IRC bot).

I've checked and it doesn't seem to affect functionality - dashboards are all correctly owned, modifying users doesn't result in incorrect values being assigned etc.

I can send a pull to change the values if you like, or have you already changed it in your branch?

**Comment by murrant** [comment]
The plan is to change the class to extend MysqlAuth (to remove code duplication) and remove extraneous functions.  I would prefer if you make the change as I can only test with the unit tests.

**Pull request merged**